R/run_eggla_gwas.R

```diff
@@ -586,61 +619,48 @@ run_eggla_gwas <- function(
-      if (use_info) {
-        output_dt <- data.table::merge.data.table(
-          y = results[j = A1 := NULL],
-          x = annot,
-          by.y = c("ID", "CHROM", "POS", "ALT", "REF"),
-          by.x = c("SNPID", "CHR", "POS", "EFFECT_ALLELE", "NON_EFFECT_ALLELE"),
-          all.y = TRUE
-        )
-        data.table::setcolorder(
-          x = output_dt,
-          neworder = c(
-            "trait_model",
-            "SNPID", "STRAND", "BUILD",
-            "CHR", "POS", "EFFECT_ALLELE", "NON_EFFECT_ALLELE",
-            "N", "N0", "N1", "N2", "EAF",
-            "BETA", "SE", "P",
-            "HWE_P", "CALL_RATE",
-            "IMPUTED", "INFO_TYPE", "INFO",
-            "T_STAT", "OBS_CT", "ERRCODE"
-          )
-        )
-      } else {
-        output_dt <- data.table::setnames(
-          x = results[j = A1 := NULL][
-            j = `:=`(
-              STRAND = strand,
-              BUILD = build
-            )
-          ],
-          old = c("ID", "CHROM", "POS", "ALT", "REF"),
-          new = c("SNPID", "CHR", "POS", "EFFECT_ALLELE", "NON_EFFECT_ALLELE")
-        )
-        data.table::setcolorder(
-          x = output_dt,
-          neworder = intersect(
-            x = c(
-              "trait_model",
-              "SNPID", "STRAND", "BUILD",
-              "CHR", "POS", "EFFECT_ALLELE", "NON_EFFECT_ALLELE",
-              "N", "N0", "N1", "N2", "EAF",
-              "BETA", "SE", "P",
-              "HWE_P", "CALL_RATE",
-              "IMPUTED", "INFO_TYPE", "INFO",
-              "T_STAT", "OBS_CT", "ERRCODE"
+      data.table::fwrite(
+        x = data.table::setnames(
+          x = Reduce(
+            f = function(x, y) merge(x, y, by = intersect(colnames(x), colnames(y))),
+            x = list(
+              data.table::fread(sprintf("%s.vmiss", results_file))[
+                j = list(
+                  `#CHROM`,
+                  ID,
+                  F_MISS,
+                  N = OBS_CT - MISSING_CT
+                )
+              ],
+              data.table::fread(sprintf("%s.afreq", results_file))[j = -c("OBS_CT")],
+              data.table::fread(sprintf("%s.hardy", results_file))[
+                j = list(
+                  `#CHROM`,
+                  ID,
+                  REF = A1, # weirdly A1 for HWE is ref in PLINK2
+                  HWE_P = P,
+                  HOM_REF_N = HOM_A1_CT,
+                  HET_REF_N = HET_A1_CT,
+                  TWO_ALT_N = TWO_AX_CT
+                )
+              ],
+              if (exists("annot")) annot else NULL
             ),
-            y = colnames(output_dt)
-          )
-        )
-      }
-      data.table::fwrite(x = output_dt, file = output_results_file)
+            init = results
+          ),
+          old = c("ID", "CHROM", "POS", "A1"),
+          new = c("SNPID", "CHR", "POS", "EFFECT_ALLELE")
+        ),
+        file = output_results_file
+      )
```

Comment:
Combine GLM results, with missing rate, Hardy-Weinberg, allele frequencies, and imputed info fields.